### PR TITLE
feature: added new module ssi-sdk-siopv2-oid4vp-rp-rest-client

### PR DESCRIPTION
**Pull request by sksadjad**
created  the new module `ssi-sdk-siopv2-oid4vp-rp-rest-client` for calling the rest endpoints

### CODE REVIEW

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts [hunk 1]
```diff
@@ -31,10 +31,10 @@ export class SiopV2OID4VpRpRestClient implements IAgentPlugin {
     }
   }
 
-  private async deleteDefinitionCorrelation(args: IDeleteDefinitionCorrelationArgs, context: IRequiredContext) {
+  private async deleteDefinitionCorrelation(args: IDeleteDefinitionCorrelationArgs, context: IRequiredContext): Promise<void> {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:34]: You are not deleting a definition or anything. You are deleting the auth session the RP has using the correlation id.

So something more appropropriate would be.

removeAuthRequestSession

Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts]: changed to `removeAuthRequestSession`

File: packages/siopv2-oid4vp-rp-rest-client/__tests__/localAgent.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+const definitionId = '9449e2db-791f-407c-b086-c21cc677d2e0'
+const baseUrl = 'https://ssi-backend.sphereon.com'
+
+const agent = createAgent<IResolver>({
+  plugins: [new SiopV2OID4VpRpRestClient(baseUrl, definitionId)],
+})
+
+describe('@sphereon/siopv2-oid4vp-rp-rest-client', () => {
+  it('should call the mock endpoint for deleteDefinitionCorrelation', async () => {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/__tests__/localAgent.test.ts:14]: Update test name

File: packages/siopv2-oid4vp-rp-rest-client/__tests__/restAgent.test.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  plugins: [new SiopV2OID4VpRpRestClient(baseUrl, definitionId)],
+})
+
+describe('@sphereon/siopv2-oid4vp-rp-rest-client', () => {
+  it('should call the endpoint for deleteDefinitionCorrelation', async () => {
+    const authRequest = await agent.generateAuthRequest({})
+    agent.removeAuthRequestSession({
+      correlationId: authRequest.correlationId,
+    })
+    await new Promise((f) => setTimeout(f, 5000))
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/__tests__/restAgent.test.ts:17]: Why wait 5 seconds?. Rather would see a shorter timeout, or a while loop with guard until it is either deleted or 5 seconds have expired

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  IRemoveAuthRequestSessionArgs,
+} from '../index'
+import Debug from 'debug'
+import { IAgentPlugin } from '@veramo/core'
+
+const debug = Debug('ssi-sdk-siopv2-oid4vp-rp-rest-client:SiopV2OID4VpRpRestClient')
+// todo merge this with Niels's branch feature/siop-verifier and use those classes/definitions
+export class SiopV2OID4VpRpRestClient implements IAgentPlugin {
+  readonly methods: ISiopV2OID4VpRpRestClient = {
+    removeAuthRequestSession: this.removeAuthRequestSession.bind(this),
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:17]: please rename these methods, as they will end up as global methods in a configured agent. See the siopv2-oid4vp-auth package for naming. Should become something like 

siopClientRemoveAuthRequestSession or something

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  private async removeAuthRequestSession(args: IRemoveAuthRequestSessionArgs, context: IRequiredContext): Promise<void> {
+    const baseUrl = this.checkBaseUrlParameter(args.baseUrl)
+    const definitionId = this.checkDefinitionIdParameter(args.definitionId)
+    fetch(this.uriWithBase(`/webapp/definitions/${definitionId}/auth-requests/${args.correlationId}`, baseUrl), {
+      method: 'DELETE',
+    })
+  }
+
+  private async getAuthStatus(args: IGetAuthStatusArgs, context: IRequiredContext): Promise<any> {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:42]: Why return an any? We have interface for the payloads. This means any implementation is left on their own to figure out what the payload is

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        correlationId: args.correlationId,
+        definitionId,
+      }),
+    })
+    debug(`auth status response: ${statueResponse}`)
+    const success = statueResponse && statueResponse.status >= 200 && statueResponse.status < 400
+    if (success) {
+      return await statueResponse.json()
+    }
+    throw Error(`Statue has returned ${statueResponse.status}`)
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:61]: First check whether the REST endpoint in my branch returns an error object or not. If that is the case return that in the error.

 also a typo in statusResponse

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+export interface IAuthStatusResponse {
+  status: AuthorizationRequestStateStatus | AuthorizationResponseStateStatus
+  correlationId: string
+  error?: string
+  definitionId: string
+  lastUpdated: number
+  payload?: AuthorizationResponsePayload // Only put in here once the status reaches Verified on the RP side
+}
+
+export declare enum AuthorizationResponseStateStatus {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts:41]: These are defined in the common package

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  CREATED = 'created',
+  SENT = 'sent',
+  RECEIVED = 'received',
+  VERIFIED = 'verified',
+  ERROR = 'error',
+}
+
+export declare enum AuthorizationRequestStateStatus {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts:49]: These are defined in the common package

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+export interface IGenerateAuthRequestURIResponse {
+  correlationId: string
+  definitionId: string
+  authRequestURI: string
+  authStatusURI: string
+  baseUrl?: string
+}
+
+export interface IAuthStatusResponse {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts:32]: These are defined in the common package

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+export interface IRemoveAuthRequestSessionArgs {
+  correlationId: string
+  baseUrl?: string
+  definitionId?: string
+}
+
+export type IGetAuthStatusArgs = IGenerateAuthRequestURIResponse
+
+export interface IGenerateAuthRequestURIResponse {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts:24]: These are defined in the common package

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+export declare enum AuthorizationRequestStateStatus {
+  CREATED = 'created',
+  SENT = 'sent',
+  RECEIVED = 'received',
+  VERIFIED = 'verified',
+  ERROR = 'error',
+}
+
+export interface AuthorizationResponsePayload {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts:57]: These are defined in the common package

**Review by nklomp — CHANGES_REQUESTED** [review]
See remarks.

File: packages/siopv2-oid4vp-rp-rest-client/__tests__/restAgent.test.ts (hunk 3 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/__tests__/restAgent.test.ts:18]: just wanted to be safe on this. in the new version (in the ssi-sdk) you're returning a response, so I will not need this anymore once that changed, I'll just `await` here

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts (hunk 4 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts]: changed to `siopClientRemoveAuthRequestSession` also changed other two method names to a more specific name

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts (hunk 5 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts]: changed to the actual response object from common

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts (hunk 6 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts]: yes you're returning it/ changed to get the response json, if any error occurs , I throw a general error here

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts (hunk 7 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts]: removed and used that one

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts (hunk 10 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts]: removed and used that one

File: packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts (hunk 11 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/types/ISiopV2OID4VpRpRestClient.ts]: removed and used that one

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  ): Promise<GenerateAuthRequestURIResponse> {
+    const baseUrl = this.checkBaseUrlParameter(args.baseUrl)
+    const definitionId = this.checkDefinitionIdParameter(args.definitionId)
+    const url = this.uriWithBase(`/webapp/definitions/${definitionId}/auth-request-uri`, baseUrl)
+    const origResponse = await fetch(url)
+    const success = origResponse && origResponse.status >= 200 && origResponse.status < 400
+    if (success) {
+      return await origResponse.json()
+    }
+    throw Error(`calling ${url} returned ${origResponse.status}`)
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:76]: Do not swallow errors at this level. This is a REST client. Whether someone using this client wants to show the actual error or create their own simpler error is up to them.

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    if (definitionId) {
+      this.definitionId = definitionId
+    }
+  }
+
+  private async siopClientRemoveAuthRequestSession(args: ISiopClientRemoveAuthRequestSessionArgs, context: IRequiredContext): Promise<void> {
+    const baseUrl = this.checkBaseUrlParameter(args.baseUrl)
+    const definitionId = this.checkDefinitionIdParameter(args.definitionId)
+    fetch(this.uriWithBase(`/webapp/definitions/${definitionId}/auth-requests/${args.correlationId}`, baseUrl), {
```
Comment (nklomp) [inline packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:37]: Isn't this returning a response? This is a client, so it should return the API responses

**Review by nklomp — CHANGES_REQUESTED** [review]
See remarks.

Just pas responses back, do not swallow exception in a REST client, that should happen in the layer that is calling into the client

File: packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts (hunk 13 above)
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/src/agent/SiopV2OID4VpRpRestClient.ts:37]: in the current code no
```
this.express.delete("/webapp/definitions/:definitionId/auth-requests/:correlationId", (request, response) => {
                const correlationId: string = request.params.correlationId
                this.verifier.sessionManager.deleteStateForCorrelationId(correlationId)
            }
        )
```

File: packages/siopv2-oid4vp-rp-rest-client/__tests__/localAgent.test.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import { createAgent, IResolver } from '@veramo/core'
+// @ts-ignore
+import nock from 'nock'
+import { SiopV2OID4VpRpRestClient } from '../src'
+
+const definitionId = '9449e2db-791f-407c-b086-c21cc677d2e0'
+const baseUrl = 'https://ssi-backend.sphereon.com'
```
[the remark this answers is not in the document]
Comment (sksadjad) [reply packages/siopv2-oid4vp-rp-rest-client/__tests__/localAgent.test.ts]: changed to `https://my-siop-endpoint`

**Review by nklomp — APPROVED** [review]

**Pull request merged**